lib/upipe-modules/upipe_dup.c

```diff
@@ -284,17 +275,14 @@ static int upipe_dup_set_flow_def(struct upipe *upipe,
         uref_free(upipe_dup->flow_def);
     upipe_dup->flow_def = flow_def_dup;
 
-    struct uchain *uchain;
-    ulist_foreach (&upipe_dup->outputs, uchain) {
-        struct upipe_dup_output *upipe_dup_output =
-            upipe_dup_output_from_uchain(uchain);
+    for (struct upipe *output = NULL;
```

Comment:
It is a question of point of view, ulist_foreach (outputs) looks pretty straightforward to me :).